federation/pkg/federatedtypes/hpa.go

```diff
@@ -0,0 +1,926 @@
+
+	remainingReplicas := replicaNums{
+		min: requestedReplicas.min - currentReplicas.min,
+		max: requestedReplicas.max - currentReplicas.max,
+	}
+
+	// Pass 2: reduction of replicas if needed ( situation that fedHpa updated replicas
+	// to lesser then existing).
+	// In this pass, we remain pessimistic and reduce one replica per cluster at a time.
+	if remainingReplicas.min < 0 {
+		excessMin := (remainingReplicas.min * int32(-1))
+		remainingReplicas.min = reduceMinReplicas(excessMin, clusterLists.availableMin, scheduleState)
+	}
+	if remainingReplicas.max < 0 {
+		excessMax := (remainingReplicas.max * int32(-1))
+		remainingReplicas.max = reduceMaxReplicas(excessMax, clusterLists.availableMax, scheduleState)
+	}
+
+	toDistribute := replicaNums{
+		min: remainingReplicas.min + int32(clusterLists.availableMin.Len()),
+		max: remainingReplicas.max + int32(clusterLists.availableMax.Len()),
+	}
+
+	// Pass 3: Distribute Max and then Min.
+	// Here we first distribute max and then (in the next loop)
+	// distribute min into those clusters which already get the
+	// max fixed.
+	// In this process we might not meet the min limit and total of
+	// min limits might remain more then the requested federated min.
+	// This is partially because a min per cluster cannot be lesser
+	// then 1, but min could be requested as 1 at federation.
+	// Additionally we first increase replicas into those clusters
+	// which already have hpa's and are in a condition to increase.
+	// This will save cluster related resources for the user, such that
+	// if an already existing cluster can satisfy users request why send
+	// the workload to another.
+	// We then go ahead to give the replicas to those which do not
+	// have any hpa. In this pass however we try to ensure that all
+	// our Max are consumed in this reconcile.
+	distributeMaxReplicas(toDistribute.max, clusterLists, rdc, currentObjs, scheduleState)
+
+	// We distribute min to those clusters which:
+	// 1 - can adjust min (our increase step would be only 1)
+	// 2 - which do not have this hpa and got max(increase step rdcMin)
+	// We might exhaust all min replicas here, with
+	// some clusters still needing them. We adjust this in finalise by
+	// assigning min replicas to 1 into those clusters which got max
+	// but min remains 0.
+	distributeMinReplicas(toDistribute.min, clusterLists, rdc, currentObjs, scheduleState)
+
+	return finaliseScheduleState(scheduleState)
+}
+
+func (a *HpaAdapter) ScheduleObject(cluster *federationapi.Cluster, clusterObj pkgruntime.Object, federationObjCopy pkgruntime.Object, schedulingInfo interface{}) (pkgruntime.Object, ScheduleAction, error) {
+	// Update federated status info
+	typedInfo := schedulingInfo.(*hpaSchedulingInfo)
+	if clusterObj != nil {
+		clusterHpa := clusterObj.(*autoscalingv1.HorizontalPodAutoscaler)
+		if clusterHpa.Status.CurrentCPUUtilizationPercentage != nil {
+			*typedInfo.fedStatus.currentCPUUtilizationPercentage +=
```

Comment:
I don't really like that you're storing a very large intermediate number, that is not actually a percentage, in a field intended to store a percentage.  And then only in a different function below do you divide the large intermediate number by the number of replicas, to arrive at a percentage, which you then overwrite the large intermediate number with.  This makes for fragile, highly coupled code (primarily because the structure returned by this function must return the large intermediate number rather than the percentage, and updateStatus() below must receive the large intermediate number rather than the percentage, in order for the code to work.

One simple suggestion to improve it would be to rename the currentCPUUtilizationPercentage field in hpaSchedulingInfo to something like aggregatedCPUUtilization, and then use it to calculate the percentage (in updateStatus below, which you could store there in a separate, local variable).

Sorry, this is a very convoluted explanation, but hopefully it makes sense :-)